org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/folding/DefaultJavaFoldingStructureProvider.java

```diff
@@ -581,16 +581,15 @@ private void createFoldingRegion(int start, int length, boolean collapse, boolea
 
 
 		private IJavaElement resolveJavaElementAt(int offset, boolean checkSurrounding) {
-			IJavaElement[] elements;
 			try {
-				elements= topLevelCompilationUnit.codeSelect(offset, 0);
+				IJavaElement element= topLevelCompilationUnit.getElementAt(offset - 1);
```

Comment:
Could you please give a hint why this change affects performance?

Reply:
Though I think that with `getElementAt` the recursion should actually not be necessary since it returns "Returns the smallest element within this Java type root that includes the given source position" - I hope I can take a look at it tomorrow.